### PR TITLE
coradoc integration

### PR DESCRIPTION
**Pull request by xyz65535**
#84
### Metanorma PR checklist

 - [ ] Breaking changes (list related PRs)
 - [ ] Documentation update required ([create task for this](https://github.com/metanorma/metanorma.org/issues/new))
 - [ ] External dependency introduced ([documentation update need](https://github.com/metanorma/metanorma.org/issues/new))
 - [ ] Gem with native library introduced

### CODE REVIEW

File: lib/reverse_adoc/converters/aside.rb [hunk 1]
```diff
@@ -1,11 +1,14 @@
 module ReverseAdoc
   module Converters
     class Aside < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {})
         id = node['id']
         anchor = id ? "[[#{id}]]\n" : ""
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/aside.rb:6]: These two lines should be removed, they are not used.

File: lib/reverse_adoc/converters/blockquote.rb [hunk 2]
```diff
@@ -1,15 +1,17 @@
 module ReverseAdoc
   module Converters
     class Blockquote < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {})
         id = node['id']
         anchor = id ? "[[#{id}]]\n" : ""
         cite = node['cite']
         attrs = cite ? "[quote, #{cite}]\n" : ""
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/blockquote.rb:8]: Please remove all fake asciidoc, e.g. `[quote, ...]`

File: lib/reverse_adoc/converters/blockquote.rb [hunk 3]
```diff
@@ -1,15 +1,17 @@
 module ReverseAdoc
   module Converters
     class Blockquote < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {})
         id = node['id']
         anchor = id ? "[[#{id}]]\n" : ""
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/blockquote.rb:6]: Please remove all fake asciidoc, e.g. `[[...]]`

File: lib/reverse_adoc/converters/br.rb [hunk 4]
```diff
@@ -1,8 +1,12 @@
 module ReverseAdoc
   module Converters
     class Br < Base
+      def to_coradoc(node, state = {})
+        Coradoc::Document::Inline::HardLineBreak.new
+      end
+
       def convert(node, state = {})
-        " +\n"
+        Coradoc::Generator.gen_adoc(to_coradoc(node, state))
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/br.rb:9]: Why do we have `to_coradoc` and `convert`?

File: lib/reverse_adoc/converters/drop.rb [hunk 5]
```diff
@@ -1,6 +1,9 @@
 module ReverseAdoc
   module Converters
     class Drop < Base
+      def to_coradoc(node, state = {})
+        convert(node, state)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/drop.rb:5]: What does this do?

File: lib/reverse_adoc/converters/figure.rb [hunk 6]
```diff
@@ -1,12 +1,14 @@
...
     class Figure < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {})
         id = node['id']
-        anchor = id ? "[[#{id}]]\n" : ""
         title = extract_title(node)
-        title = ".#{title}\n" unless title.empty?
-        "\n\n#{anchor}#{title}====\n" << treat_children(node, state).strip << "\n====\n\n"
+        content = treat_children(node, state).strip
+        Coradoc::Document::Block.new(title, lines: content.lines, type: :example, id: id)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/figure.rb:8]: The `content.lines` should be converted to Coradoc objects before putting in.

e.g.
```html
<figure>
  <img src="pic_trulli.jpg" alt="Trulli" style="width:100%">
  <figcaption>Fig.1 - Trulli, Puglia, Italy.</figcaption>
</figure>
```

File: lib/reverse_adoc/converters/h.rb [hunk 7]
```diff
@@ -1,14 +1,19 @@
 module ReverseAdoc
   module Converters
     class H < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {})
         id = node['id']
         anchor = id ? "[[#{id}]]" : ""
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/h.rb:6]: No fake asciidoc please.

File: lib/reverse_adoc/converters/h.rb [hunk 8]
```diff
@@ -1,14 +1,19 @@
...
         internal_anchor = treat_children_anchors(node, state) || ""
         anchor.empty? and anchor = internal_anchor
         anchor.empty? or anchor += "\n"
-        prefix = '=' * (node.name[/\d/].to_i + 1)
-        ["\n", anchor, prefix, ' ', treat_children_no_anchors(node, state), "\n"].join
+        level_str = '=' * (node.name[/\d/].to_i + 1)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/h.rb:10]: Count the levels using an integer, not like this.

File: lib/reverse_adoc/converters/img.rb [hunk 9]
```diff
@@ -55,26 +55,25 @@ def copy_temp_file(imgdata)
...
-        attrs = alt
-        attrs = "\"\"" if (width || height) && alt.nil?
-        attrs += ",#{width}" if width
-        attrs += ",#{height}" if width && height
-        [anchor, title, "image::", src, "[", attrs, "]"].join("")
+        Coradoc::Document::Inline::Image.new(id: id,
+          title: title, src: src, alt: alt, width: width, height: height)
+      end
+      def convert(node, state = {})
+        Coradoc::Generator.gen_adoc(to_coradoc(node, state))
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/img.rb:76]: Why isn't it just `node.to_adoc`?

File: lib/reverse_adoc/converters/li.rb [hunk 10]
```diff
@@ -1,12 +1,18 @@
...
   module Converters
     class Li < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {})
+        # convert(node, state)
         id = node['id']
         anchor = id ? "[[#{id}]]" : ""
-        content     = treat_children(node, state)
+        content     = treat_children_coradoc(node, state)
         prefix      = prefix_for(node, state)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/li.rb:9]: This line is not used.

File: lib/reverse_adoc/converters/li.rb [hunk 11]
```diff
@@ -1,12 +1,18 @@
 module ReverseAdoc
   module Converters
     class Li < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {})
+        # convert(node, state)
         id = node['id']
         anchor = id ? "[[#{id}]]" : ""
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/li.rb:7]: No fake asciidoc please.

File: lib/reverse_adoc/converters/ol.rb [hunk 12]
```diff
@@ -1,12 +1,36 @@
 module ReverseAdoc
   module Converters
     class Ol < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {}) #FIXIT
+        # convert(node, state)
         id = node['id']
         anchor = id ? "[[#{id}]]\n" : ""
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/ol.rb:7]: No fake asciidoc please.

File: lib/reverse_adoc/converters/ol.rb [hunk 13]
```diff
@@ -1,12 +1,36 @@
...
-        "\n#{anchor}#{attrs}" << treat_children(node, state.merge(ol_count: ol_count))
+        items = treat_children_coradoc(node, state.merge(ol_count: ol_count))
+        options = {}
+        options[:id] = id
+        options[:anchor] = anchor
+        options[:ol_count] = ol_count
+        options[:attrs] = attrs
+        list_type = get_list_type(node, state)
+        if list_type == :ordered
+          Coradoc::Document::List.new(items, options)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/ol.rb:18]: Use List::Ordered.new for clarity.

File: lib/reverse_adoc/converters/p.rb [hunk 14]
```diff
@@ -1,14 +1,16 @@
...
-        anchor = id ? "[[#{id}]]\n" : ""
-        if state[:tdsinglepara]
-          "#{anchor}" << treat_children(node, state).strip
-        else
-          "\n\n#{anchor}" << treat_children(node, state).strip << "\n\n"
-        end
+        content = treat_children(node, state).strip
+        options = {}
+        options[:id] = id if id
+        options[:tdsinglepara] = true if state[:tdsinglepara]
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/p.rb:9]: I don't know what this is... let's delete it?

File: lib/reverse_adoc/converters/pre.rb [hunk 15]
```diff
@@ -1,17 +1,19 @@
...
     class Pre < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {})
         id = node['id']
-        anchor = id ? "[[#{id}]]\n" : ""
         lang = language(node)
         content = treat_children(node, state)
         if lang
-          "\n\n#{anchor}[source,#{lang}]\n----\n" << content.lines.to_a.join("") << "\n----\n\n"
+          Coradoc::Document::Block.new(nil, lines: content.lines, delimiter: '----', lang: lang, id: id)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/pre.rb:9]: `<pre>` should lead:
* a `Literal` block
* if it has a `lang` attribute, make it a `SourceCode` block.

There's no need to give any delimiter for either.

File: lib/reverse_adoc/converters/table.rb [hunk 16]
```diff
@@ -1,13 +1,18 @@
 module ReverseAdoc
   module Converters
     class Table < Base
-      def convert(node, state = {})
+      def to_coradoc(node, state = {}) #FIXIT
+        # convert(node,state) #PLACEHOLDER
         id = node['id']
-        anchor = id ? "[[#{id}]]\n" : ""
         title = extract_title(node)
         title = ".#{title}\n" unless title.empty?
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/table.rb:8]: Remove this line.

File: lib/reverse_adoc/converters/td.rb [hunk 17]
```diff
@@ -13,8 +13,18 @@ def convert(node, state = {})
           node.at(".//p") && !singlepara
         style = "a" if adoccell
         delim = adoccell ? "\n" : " "
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/td.rb:15]: Delete all lines that are unused.

File: lib/reverse_adoc/html_converter.rb [hunk 18]
```diff
@@ -46,11 +46,13 @@ def self.convert(input, options = {})
...
       ReverseAdoc.config.with(options) do
-        result = ReverseAdoc::Converters.lookup(root.name).convert(root)
-        ReverseAdoc.cleaner.tidy(result)
+        result = ReverseAdoc::Converters.lookup(root.name).to_coradoc(root)
       end
     end
+    def self.convert(input, options = {})
+      result = Coradoc::Generator.gen_adoc(to_coradoc(input, options))
+      ReverseAdoc.cleaner.tidy(result)
+    end
```
Comment (ronaldtse) [inline lib/reverse_adoc/html_converter.rb:56]: Why do we need any cleaning after putting it into a `Coradoc::Document`? The to_adoc results should be perfect.

**Review by ronaldtse — CHANGES_REQUESTED** [review]

File: lib/reverse_adoc/converters/a.rb [hunk 19]
```diff
@@ -13,7 +13,22 @@ def to_coradoc(node, state ={})
...
 
-        Coradoc::Document::Inline::Anchor.new(id: id, href: href, title: title, name: name)
+        if /^_Toc\d+$|^_GoBack$/.match id
+          ""
+        elsif !id.nil? && !id.empty?
+          Coradoc::Document::Inline::Anchor.new(id)
+        elsif href.to_s.start_with?('#')
+          href = href.sub(/^#/, "").gsub(/\s/, "").gsub(/__+/, "_")
+          if name.to_s.empty?
+            Coradoc::Document::Inline::CrossReference.new(href)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/a.rb:23]: A cross reference links an optional custom text to a target. The target can be an anchor in the document, or an anchor in a separate document.

Comment (ronaldtse) [reply lib/reverse_adoc/converters/a.rb:23]: Maybe combine this case with the next case into an initialize method that takes parameters.

File: lib/reverse_adoc/converters/aside.rb [hunk 20]
```diff
@@ -2,10 +2,8 @@ module ReverseAdoc
   module Converters
     class Aside < Base
       def to_coradoc(node, state = {})
-        id = node['id']
-        anchor = id ? "[[#{id}]]\n" : ""
-        content = treat_children(node, state)#.strip
-        Coradoc::Document::Block.new(nil, lines: content.lines, type: :side)
+        content = treat_children(node, state)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/aside.rb:5]: Can we call “treat_children” just “parse” on the children?

File: lib/reverse_adoc/converters/blockquote.rb [hunk 21]
```diff
@@ -3,13 +3,13 @@ module Converters
...
       def to_coradoc(node, state = {})
         id = node['id']
-        anchor = id ? "[[#{id}]]\n" : ""
         cite = node['cite']
-        attrs = cite ? "[quote, #{cite}]\n" : ""
+        attributes = cite.nil? ? nil :  Coradoc::Document::AttributeList.new("quote", cite)
         content = treat_children(node, state).strip
         content = ReverseAdoc.cleaner.remove_newlines(content)
-        Coradoc::Document::Block.new(nil, lines: content.lines, type: :quote, attributes: attrs)
+        Coradoc::Document::Block::Quote.new(nil, lines: content, attributes: attributes)
```
Comment (ronaldtse) [inline lib/reverse_adoc/converters/blockquote.rb:10]: What’s the first nil argument? It looks weird.

Comment (xyz65535) [reply lib/reverse_adoc/converters/blockquote.rb:10]: It's title, previous reverse_adoc code was inconsistent with regards to blocks, I will look into how exactly are different types of asciidoc blocks converted into html

**Comment by ronaldtse** [comment]
@xyz65535 once you're done please help work on this:
* https://github.com/metanorma/reverse_adoc/issues/90

Thanks!

File: lib/reverse_adoc/converters/blockquote.rb (hunk 21 above)
Comment (ronaldtse) [reply lib/reverse_adoc/converters/blockquote.rb:10]: So there are actually 3 types of HTML source:

1. Normal HTML
2. Source generated by Metanorma
  * Source: https://github.com/metanorma/mn-samples-iso
  * HTML: https://metanorma.github.io/mn-samples-iso/
3. Source generated by Word
  * (@opoudjis what are the examples?)

**Pull request merged**